backend/logic/courses/courses.js

```diff
@@ -1,19 +1,72 @@
-const mysql = require('../../sql/connection');
+const mysql = require("../../sql/connection");
+const format = require("../../validation/format")
 
-async function queryCourseByTag(tag) {
-    if (!tag) {
-        throw Error("Undefined tag")
-    }
+var queryCourseByTag = async function queryCourseByTag(tag) {
+  if (!tag) {
+    throw Error("Undefined tag")
+  }
+
+  let connection = await mysql.getNewConnection();
+  try {
+    let courses = await connection.query("SELECT course_code FROM course_tag WHERE course_tag.tag_name LIKE ?", tag);
+    connection.release();
+    return (JSON.parse(JSON.stringify(courses)));
+  } catch (error) {
+    connection.release();
+    console.error(error);
+    throw Error(error.message);
+  }
+}
+
+var addCompletedCourses = async (studentId, courses, section) => {
+  if (!format.verifyStudentId(studentId)) {
+    throw new Error("invalid format id");
+  }
+  await format.verifyCompletedCourses(courses);
+
+  let connection;
+  try {
+    connection = await mysql.getNewConnection();
+  } catch (error) {
+    throw new Error("failed to establish connection with database");
```

Comment:
Here the getNewConnection method will already throw an error "failed to connect to database if anything goes wrong, so i think you can let this get picked up by the top level error handler

Reply:
Changed